### PR TITLE
#167165057 User Can Sign Out

### PR DESCRIPTION
**Pull request by tolulope-od**
### What does this PR do?

Creates functionality for users to sign out
Creates an authentication middleware
Creates a server response module
Add pretest script
Changes blacklist table token column data type
Adds database service to Travis CI

#### Description of Task to be completed?
When a user signs out of the app (drops token), then the user should not be able to access any route protected by the token.

User should be able to sign out via the route: `api/v1/auth/logout`
 
#### How should this be manually tested?
After cloning this repo, `cd` into it and do the following:

```bash
# fetch this branch
$ git fetch origin ft-user-sign-out-167165057

# install dependencies
$ npm install

# run unit tests
$ npm test

# start the development server
$ npm run server
```

When the server is running, using Postman, make a `POST` request to the route stated above

#### Any background context you want to provide?
Given the fact that when a JSON Web Token is generated, it cannot be unsigned, the way this feature was implemented was to have a table that stores blacklisted tokens. This table is populated when a user hits the logout route, hence this route is protected by an Authentication middleware that checks for a token.

#### What are the relevant pivotal tracker stories?
[#167165057](https://www.pivotaltracker.com/story/show/167165057)

#### Screenshots (if appropriate)
*Successful logout request*

Screenshot 2019-07-25 at 16 08 01

### CODE REVIEW

**Review by OdunayoOkebunmi — COMMENTED** [review]
Kindly include screenshots.

File: controllers/AuthController.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * @export
+ * @class AuthController
+ */
+export default class AuthController {
+  /**
+   *
+   * @static
+   * @param {object} req express request object
+   * @param {object} res express response object
+   * @returns {object} returns user data
```
Comment (OdunayoOkebunmi) [inline controllers/AuthController.js:22]: kindly separate your JSDoc block tags

**Comment by tolulope-od** [comment]
> Kindly include screenshots.

Done.

File: controllers/AuthController.js (hunk 1 above)
Comment (tolulope-od) [reply controllers/AuthController.js:22]: I pulled in files from someone else's branch and must have missed this but I will make these changes. Thank you

**Review by mekzy-o — APPROVED** [review]
Good Job Tolu. Though your file changes were much. Keep it less next time.

**Review by mekzy-o — APPROVED** [review]

File: .env [hunk 2]
```diff
@@ -0,0 +1,3 @@
+DATABASE_URL=postgres://tolulopeodueke@localhost:5432/mazus_ah_backend
+TEST_DATABASE_URL=postgres://tolulopeodueke@localhost:5432/mazus_ah_backend_test
+SECRET_KEY=master
```
Comment (mazma1) [inline .env]: What is `.env` doing on Github again???

File: database/models/user.js [hunk 3]
```diff
@@ -3,92 +3,95 @@
...
-  }, {});
+    {}
+  );
   User.associate = function(models) {
     User.hasOne(models.Profile, {
       foreignKey: 'userId',
       as: 'profile',
       onDelete: 'CASCADE',
-      onUpdate: 'CASCADE',
+      onUpdate: 'CASCADE'
```
Comment (mazma1) [inline database/models/user.js]: Please put back the trailing commas you removed from this file as they are important.

Note: Don't go making changes to files that have nothing to do with what you are working on.

File: .env (hunk 2 above)
Comment (tolulope-od) [reply .env]: I didn't know this was added again, I'll remove it. Thanks

File: controllers/AuthController.js (hunk 1 above)
Comment (tolulope-od) [reply controllers/AuthController.js:22]: Done

File: .env (hunk 2 above)
Comment (tolulope-od) [reply .env]: Done

File: database/models/user.js (hunk 3 above)
Comment (tolulope-od) [reply database/models/user.js]: Done

**Review by Tunji17 — COMMENTED** [review]
Good job @tolulope-od but is there a reason why the normal migrate script is crossing env to test before running migrate?

**Comment by tolulope-od** [comment]
> Good job @tolulope-od but is there a reason why the normal migrate script is crossing env to test before running migrate?

Thanks for pointing that out. That was an oversight. I'll fix that now

File: database/migrations/20190723105203-create-user.js [hunk 4]
```diff
@@ -45,7 +45,7 @@ module.exports = {
       allowNull: false,
       type: Sequelize.DATE,
       defaultValue: Sequelize.literal('CURRENT_TIMESTAMP'),
-    }
+    },
   }),
-  down: (queryInterface, Sequelize) => queryInterface.dropTable('Users')
+  down: (queryInterface, Sequelize) => queryInterface.dropTable('Users'),
```
Comment (mazma1) [inline database/migrations/20190723105203-create-user.js:50]: Please enforce trailing commas for only multiline array and object statements. It doesn't apply here

Comment (tolulope-od) [reply database/migrations/20190723105203-create-user.js:50]: ESLint flags it when I do that.

**Pull request closed without merging**